mqtt/mqtt-broker/src/persist.rs

```diff
@@ -136,6 +145,211 @@ impl FileFormat for BincodeFormat {
     }
 }
 
+#[derive(Clone, Debug, Default)]
+pub struct ConsolidatedStateFormat;
+
+impl ConsolidatedStateFormat {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl FileFormat for ConsolidatedStateFormat {
+    type Error = Error;
+
+    fn load<R: Read>(&self, reader: R) -> Result<BrokerState, Self::Error> {
+        let decoder = GzDecoder::new(reader);
+        fail_point!("bincodeformat.load.deserialize_from", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        let state = bincode::deserialize_from(decoder)
+            .context(ErrorKind::Persist(ErrorReason::Deserialize))?;
+
+        let state = Resolver::resolve_state(state);
+        Ok(state)
+    }
+
+    fn store<W: Write>(&self, writer: W, state: BrokerState) -> Result<(), Self::Error> {
+        let state = Consolidator::consolidate_state(state);
+
+        let encoder = GzEncoder::new(writer, Compression::default());
+        fail_point!("bincodeformat.store.serialize_into", |_| {
+            Err(Error::from(ErrorKind::Persist(ErrorReason::Deserialize)))
+        });
+        bincode::serialize_into(encoder, &state)
+            .context(ErrorKind::Persist(ErrorReason::Serialize))?;
+        Ok(())
+    }
+}
+
+struct Consolidator {
+    payloads: HashMap<Bytes, u64>,
+    curr_id: u64,
+}
+
+impl Consolidator {
+    fn consolidate_state(state: BrokerState) -> ConsolidatedState {
+        let mut this = Self {
+            payloads: HashMap::new(),
+            curr_id: 0,
+        };
+
+        let (retained, sessions) = state.into_parts();
+
+        let retained = retained
+            .into_iter()
+            .map(|(k, v)| (k, this.consolidate_publication(v)));
+        let retained = HashMap::from_iter(retained);
+
+        let sessions = sessions
+            .into_iter()
+            .map(|s| this.consolidate_session(s))
+            .collect();
+
+        let payloads = this.payloads.into_iter().map(|(k, v)| (v, k));
+        let payloads = HashMap::from_iter(payloads);
```

Comment:
did not know that Rust allows you to shadow like that :) took me a while to realize that you actually flipping key/value

Reply:
yea, its actually really common to shadow in rust